<?xml version="1.0"?>
<launch>
  <!-- bringup robot -->
  <arg name="robot_name" default="snake" />
  <arg name="rm" default="True" />
  <arg name="sim" default="False" />
  <arg name="estimate_mode"  default= "0" />
  <arg name="headless" default="True" />
  <arg name="worldtype" default="$(find aerial_robot_simulation)/gazebo_model/world/0066.world" />
  <!-- <arg name="launch_gazebo" default="True" /> -->
  <arg name="robot_spawn_x" default="0.0"/>
  <arg name="robot_spawn_y" default="0.0"/>
  <arg name="robot_spawn_z" default="0.0"/>
  <arg name="robot_spawn_yaw" default="0.0"/>

  <!-- 追加：手の位置のソース(marker or mocap) -->
  <arg name="hand_source" default="mocap"/> 

  <!-- 追加：Gazeboからmocap疑似生成を使うかどうか -->
  <arg name="use_gazebo_mocap" default="False"/>  

  <include file="$(eval find(arg('robot_name')) + '/launch/bringup.launch')" >
    <arg name="real_machine" value="$(arg rm)" />
    <arg name="simulation" value="$(arg sim)" />
    <arg name="estimate_mode"  value= "$(arg estimate_mode)" />
    <arg name="headless" value="$(arg headless)" />
    <arg name="worldtype" value="$(arg worldtype)" />
    <!-- ここでlaunch_gazebo = simに固定 -->
    <arg name="launch_gazebo" value="$(arg sim)" />
    <arg name="spawn_x" value="$(arg robot_spawn_x)"/>
    <arg name="spawn_y" value="$(arg robot_spawn_y)"/>
    <arg name="spawn_z" value="$(arg robot_spawn_z)"/>
    <arg name="spawn_yaw" value="$(arg robot_spawn_yaw)"/>
    <arg name="demo" value="False" />
  </include>

  <!-- spawn eye camera -->
  <arg name="eye_spawn_x" default="-2.0"/>
  <arg name="eye_spawn_y" default="0.0"/>
  <arg name="eye_spawn_z" default="1.0"/>
  <arg name="eye_spawn_yaw" default="0.0"/>
  <node name="eye_camera_spawner" pkg="gazebo_ros" type="spawn_model" respawn="false" output="screen"
        args="-file $(find aerial_handover)/urdf/eye_camera.urdf.xacro -urdf -model eye -x $(arg eye_spawn_x) -y $(arg eye_spawn_y) -z $(arg eye_spawn_z) -Y $(arg eye_spawn_yaw)" if="$(arg sim)"/>

  <node pkg="image_view" type="image_view" name="image_view" args="image:=/eye/image_raw"/>
  <!-- <node pkg="image_view" type="image_view" name="image_view_left" -->
  <!-- args="image:=/eye_left/image_raw" /> -->

  <!-- <node pkg="image_view" type="image_view" name="image_view_right" -->
  <!-- args="image:=/eye_right/image_raw" /> -->



  <!-- convert mocap data to gazebo world -->
  <!-- <node name="mocap_convert" pkg="aerial_handover" type="mocap_convert.py" respawn="false" output="screen"> -->
  <!--   <remap from="hand_pose" to ="/wrist/mocap/pose"/> -->
  <!--   <remap from="eye_pose" to ="/eye/mocap/pose"/> -->
  <!-- </node> -->

  <!-- convert gazebo world to mocap data-->
  <!-- 変更：simのとき常にではなく、use_gazebo_mocapがTrueのときだけ -->
  <node name="gazebo_state_converter" pkg="aerial_handover" type="gazebo_state_converter.py" respawn="false">
    <remap from="hand_pose" to ="/wrist/mocap/pose"/>
    <remap from="eye_pose" to ="/eye/mocap/pose"/>
    </node>


  <!-- mocap -->
  <!-- 変更：hand_source == 'mocap'のときだけ起動（sim/実機どちらでも） -->
  <node pkg="mocap_optitrack" type="mocap_node" name="mocap_node" respawn="false" if="$(eval arg('hand_source') == 'mocap')">
    <rosparam subst_value="true">
      rigid_bodies:
         '2':
               pose: /hand_pose
               pose2d: /wrist/mocap/ground_pose
               child_frame_id: baselink
               parent_frame_id: world
         '3':
               pose: /eye/mocap/pose
               pose2d: /eye/mocap/ground_pose
               child_frame_id: baselink
               parent_frame_id: world
      optitrack_config:
         multicast_address: 239.255.42.99
         command_port: 1510
         data_port: 1511
    </rosparam>
  </node>

  <node pkg="aerial_handover" type="interactive_marker.py" name="interactive_marker" output="screen" if="$(eval arg('hand_source') == 'marker')">
    <param name="frame_id" value="world" />
    <remap from="/desired_3D_pose" to="/hand_pose" />
  </node>

  <!-- <node pkg="aerial_handover" type="set_hand_pose_gazebo.py" name="set_hand_pose_gazebo" output="screen" if="$(arg sim)"/> -->

  <!-- spawn hand model -->
  <arg name="bottle_spawn_x" default="-0.2"/>
  <arg name="bottle_spawn_y" default="0.0"/>
  <arg name="bottle_spawn_z" default="0.0"/>
  <arg name="bottle_spawn_yaw" default="0.0"/>
  <node name="bottle_spawner" pkg="gazebo_ros" type="spawn_model" respawn="false" output="screen"
        args="-file $(find aerial_handover)/urdf/bottle.urdf.xacro -urdf -model bottle -x $(arg bottle_spawn_x) -y $(arg bottle_spawn_y) -z $(arg bottle_spawn_z) -Y $(arg bottle_spawn_yaw)" if="$(arg sim)"/>
  <node pkg="aerial_handover" type="pub_obj_pose_gazebo.py" name="pub_obj_pose_gazebo" output="screen" if="$(arg sim)"/>
  <node pkg="aerial_handover" type="hand_grasp_snake_bite.py" name="hand_grasp_snake_bite" output="screen" if="$(arg sim)"/>

  <!-- spawn hand model -->
  <arg name="hand_spawn_x" default="2.0"/>
  <arg name="hand_spawn_y" default="0.0"/>
  <arg name="hand_spawn_z" default="1.0"/>
  <arg name="hand_spawn_yaw" default="0.0"/>
  
  <param name="/hand/robot_description"  
         textfile="$(find aerial_handover)/urdf/hand.urdf.xacro" />  
  <node name="hand_spawner" pkg="gazebo_ros" type="spawn_model" respawn="false" output="screen"
        args="-file $(find aerial_handover)/urdf/hand.urdf.xacro -urdf -model hand -x $(arg hand_spawn_x) -y $(arg hand_spawn_y) -z $(arg hand_spawn_z) -Y $(arg hand_spawn_yaw)" if="$(arg sim)"/>

  <include file="$(find ros_tcp_endpoint)/launch/endpoint.launch" >
  </include>
  
  
</launch>
